File: Second Time/C++/src/4_Median_of_Two_Sorted_Arrays/4_Median_of_Two_Sorted_Arrays.cpp
/*
 * @Author: fieryheart 
 * @Date: 2019-11-22 19:33:02 
 * @Last Modified by: fiery
 * @Last Modified time: 2019-11-22 20:27:30
 */

#include "../../lib/Root.h"

class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int len1 = nums1.size(), len2 = nums2.size();
        vector<int> vec1 = nums1;
        vector<int> vec2 = nums2;
        if(len1 > len2) {
            vec1 = nums2;
            vec2 = nums1;
        }
        len1 = vec1.size();
        len2 = vec2.size();
        int vec1_left =  0, vec1_right = len1;
        while(true) {
            int i = (vec1_left + vec1_right) / 2, j = (len1 + len2 + 1) / 2 - i;
            if(i > 0 && vec1[i-1] > vec2[j]) {
                vec1_right = i-1;
            } else if(i < len1 && vec2[j-1] > vec1[i]) {
                vec1_left = i+1;
            } else {
                int max_of_left, min_of_right;
                if(i == 0) max_of_left = vec2[j-1];
                else if(j == 0) max_of_left = vec1[i-1];
                else max_of_left = max(vec1[i-1], vec2[j-1]);

                if((len1+len2)&1) return max_of_left / 1.0;

                if(i == len1) min_of_right = vec2[j];
                else if(j == len2) min_of_right = vec1[i];
                else min_of_right = min(vec1[i], vec2[j]);

                return (max_of_left + min_of_right) / 2.0;
            }
        }
    }
};

int main()
{
    vector<int> nums1 = {1,2};
    vector<int> nums2 = {3,4};
    Solution solu;
    double ans = solu.findMedianSortedArrays(nums1, nums2);
    Print::print(ans);
    return 0;
}

File: Second Time/C++/src/4_Median_of_Two_Sorted_Arrays/4.Median-of-Two-Sorted-Arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        if not len(nums1): return (nums2[(len(nums2)-1)//2]+nums2[len(nums2)//2]) / 2.0
        if not len(nums2): return (nums1[(len(nums1)-1)//2]+nums1[len(nums1)//2]) / 2.0
        if len(nums1) > len(nums2):
            temp = nums1
            nums1 = nums2
            nums2 = temp
        nums1_left = 0 
        nums1_right = len(nums1)
        while True:
            i = (nums1_left + nums1_right) // 2
            j = (len(nums1)+len(nums2)+1) // 2 - i
            if i > 0 and nums1[i-1] > nums2[j]:
                nums1_right = i-1
            elif i < len(nums1) and nums2[j-1] > nums1[i]:
                nums1_left = i+1
            else:
                max_of_left = min_of_right = 0
                if i == 0: max_of_left = nums2[j-1]
                elif j == 0: max_of_left = nums1[i-1]
                else: max_of_left = max(nums1[i-1], nums2[j-1])
                
                if (len(nums1)+len(nums2))&1: return max_of_left / 1.0
                
                if i == len(nums1): min_of_right = nums2[j]
                elif j == len(nums2): min_of_right = nums1[i]
                else: min_of_right = min(nums1[i], nums2[j])
                    
                return (max_of_left+min_of_right) / 2.0
                
        